_posts/2023-05-17-Scaling_part_2.md

```diff
@@ -0,0 +1,642 @@
+In this cluster we would generate BareMetalHosts and create Clusters, Metal3Clusters, etc to benchmark the controllers.
+To give them a realistic test, we also need to fake the workload cluster API's.
+These will run separately to avoid interfering with the test (e.g. by using up all the resources in the management cluster).
+Here is a diagram that describes the setup:
+
+![diagram of test setup](/assets/2023-05-17-Scaling_part_2/scaling-fake-clusters.drawio.png)
+
+How are we going to fake the workload cluster API's then?
+The most obvious solution is to just run the real deal, i.e. the `kube-apiserver`.
+This is what would be run in a real workload cluster, together with the other components that makes up the Kubernetes control plane.
+
+## Running the Kubernetes API server
+
+There are many misconceptions, maybe even superstition, about the Kubernetes control plane.
+The fact is that it is in no way special.
+It consists of a few programs that can be run in any way you want: in a container, as a systemd unit or directly executed at the command line.
+They can run on a Node or outside of the cluster.
+You can even run multiple instances on the same host as long as you avoid port collisions.
+
+For our purposes we basically want to run as little as possible of the control plane components.
+We just need the API to be available and possible for us to populate with data that the controllers expect to be there.
+In other words, we need the API server and etcd.
+The scheduler is not necessary since we won't run any actual workload (we are just pretending the Nodes are there anyway) and the controller manager would just get in the way when we want to fake resources.
+It would, for example, try to update the status of the (fake) Nodes that we want to create.
+
+The API server will need an etcd instance to connect to.
+It will also need some TLS configuration, both for connecting to etcd and for handling service accounts.
+One simple way to generate the needed certificates is to use kubeadm.
+But before we get there we need to think about how the configuration should look like.
+
+For simplicity, we will simply run the API server and etcd in a kind cluster for now.
+It would then be easy to run them in some other Kubernetes cluster later if needed.
+Let's create it right away:
+
+```bash
+kind create cluster
+```
+
+To try to cut down on the resources required, we will also use a single multi-tenant etcd instance instead of one per API server.
+We can rely on the internal service discovery so the API server can find etcd via an address like `etcd-server.etd-system.svc.cluster.local`, instead of using IP addresses.
+Finally, we will need an endpoint where the API is exposed to the cluster where the controllers are running, but for now we can focus on just getting it up and running with `127.0.0.1:6443` as the endpoint.
+
+Based on the above, we can create a `kubeadm-config.yaml` file like this:
```

Comment:
This will be later used from the path `/tmp` would be nice to set that it should be created in the `tmp` folder